File: windows/net/samples/capture_ds_video_audio/VideoCapturePropertiesForm.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;
using DirectShowLib;
using System.Runtime.InteropServices;
using System.Runtime.InteropServices.ComTypes;

namespace CaptureDS
{
    public partial class VideoCapturePropertiesForm : Form
    {
        public VideoCapturePropertiesForm()
        {
            InitializeComponent();
        }

        void SetFormat(int formatIndex, int frameRate)
        {
            int capsCount, capSize;
            int hr = VideoConfig.GetNumberOfCapabilities(out capsCount, out capSize);
            DsError.ThrowExceptionForHR(hr);

            IntPtr pSC = Marshal.AllocHGlobal(capSize);
            AMMediaType mt = null;

            try
            {
                VideoInfoHeader vih = new VideoInfoHeader();

                hr = VideoConfig.GetStreamCaps(formatIndex, out mt, pSC);
                DsError.ThrowExceptionForHR(hr);

                if(frameRate > 0)
                {
                    Marshal.PtrToStructure(mt.formatPtr, vih);
                    vih.AvgTimePerFrame = (long)(10000000.0 / frameRate);
                    Marshal.StructureToPtr(vih, mt.formatPtr, false);
                }

                hr = VideoConfig.SetFormat(mt);
                DsError.ThrowExceptionForHR(hr);
            }
            finally
            {
                DsUtils.FreeAMMediaType(mt);
                Marshal.FreeHGlobal(pSC);
            }
        }

        private void buttonOK_Click(object sender, EventArgs e)
        {
            ComboboxItem itemFormat = comboBoxFormats.SelectedItem as ComboboxItem;
            if (itemFormat != null)
            {
                int frameRate = -1;
                ComboboxItem itemFPS = comboBoxFrameRate.SelectedItem as ComboboxItem;
                if (null != itemFPS)
                {
                    frameRate = itemFPS.Value;
                }

                try
                {
                    SetFormat(itemFormat.Value, frameRate);
                }
                catch (Exception)
                {
                    MessageBox.Show("Failed to set the selected video format.");
                }
            }

            DialogResult = DialogResult.OK;
        }

        private void buttonCancel_Click(object sender, EventArgs e)
        {
            DialogResult = DialogResult.Cancel;
        }

        public IAMStreamConfig VideoConfig { get; set; }

        struct FormatEntry
        {
            public Guid videoSubType;
            public string Name;

            public FormatEntry(Guid videoSubType, string name)
            {
                this.videoSubType = videoSubType;
                this.Name = name;
            }
        };

        private static FormatEntry[] FormatsTab =
        {
            new FormatEntry(MediaSubType.MJPG,	"MJPG"),

            new FormatEntry(MediaSubType.RGB24,	"RGB24"),
            new FormatEntry(MediaSubType.ARGB32, "ARGB32"),
            new FormatEntry(MediaSubType.RGB32,	"RGB32"),
            new FormatEntry(MediaSubType.RGB565, "RGB565"),
            new FormatEntry(MediaSubType.ARGB1555, "ARGB1555"),
            new FormatEntry(MediaSubType.RGB555, "RGB555"),
            new FormatEntry(MediaSubType.ARGB4444, "ARGB4444"),

            new FormatEntry(MediaSubType.YV12, "YV12"),
            new FormatEntry(MediaSubType.I420, "I420"),
            new FormatEntry(MediaSubType.IYUV, "IYUV"),
            new FormatEntry(MediaSubType.YUY2, "YUY2"),

            new FormatEntry(MediaSubType.NV12, "NV12"),
            new FormatEntry(MediaSubType.UYVY, "UYVY"),
            new FormatEntry(MediaSubType.Y411, "Y411"),
            new FormatEntry(MediaSubType.Y41P, "Y41P"),
            new FormatEntry(MediaSubType.YVU9, "YVU9"),
        };

        string GetSubtypeString(Guid videoSubtype)
        {
            for (int i = 0; i < FormatsTab.Length; i++)
            {
                if (FormatsTab[i].videoSubType == videoSubtype)
                    return FormatsTab[i].Name;
            }

            return null;
        }

        public class ComboboxItem
        {
            public string Text { get; set; }
            public int Value { get; set; }

            public override string ToString()
            {
                return Text;
            }
        }

        private void VideoCapturePropertiesForm_Load(object sender, EventArgs e)
        {
            try
            {
                InitFormatsList();
            }
            catch (Exception)
            {
                MessageBox.Show("Failed to set init the formats list.");
            }
        }

        private void InitFormatsList()
        {
            int hr = 0;

            if (null == VideoConfig)
                throw new Exception("VideoConfig not set");

            // enumerate the capabilities of the video capture device
            int capsCount, capSize;
            hr = VideoConfig.GetNumberOfCapabilities(out capsCount, out capSize);
            DsError.ThrowExceptionForHR(hr);

            VideoInfoHeader vih = new VideoInfoHeader();
            VideoStreamConfigCaps vsc = new VideoStreamConfigCaps();
            IntPtr pSC = Marshal.AllocHGlobal(capSize);

            try
            {
                int videoFormatIndex = -1;
                int minFps = -1;
                int maxFps = -1;

                int currentWidth = 0;
                int currentHeight = 0;
                Guid currentSubType;
                int currentFps = 0;

                {
                    AMMediaType mt = null;
                    hr = VideoConfig.GetFormat(out mt);
                    Marshal.PtrToStructure(mt.formatPtr, vih);
                    DsError.ThrowExceptionForHR(hr);

                    currentFps = (int)(10000000.0 / vih.AvgTimePerFrame);
                    currentWidth = vih.BmiHeader.Width;
                    currentHeight = vih.BmiHeader.Height;
                    currentSubType = mt.subType;

                    DsUtils.FreeAMMediaType(mt);
                }

                for (int i = 0; i < capsCount; ++i)
                {
                    AMMediaType mt = null;

                    // the video format is described in AMMediaType and VideoStreamConfigCaps
                    hr = VideoConfig.GetStreamCaps(i, out mt, pSC);
                    DsError.ThrowExceptionForHR(hr);

                    if (mt.formatType == DirectShowLib.FormatType.VideoInfo)
                    {
                        string formatName = GetSubtypeString(mt.subType);

                        if (!string.IsNullOrEmpty(formatName))
                        {
                            // copy the unmanaged structures to managed in order to check the format
                            Marshal.PtrToStructure(mt.formatPtr, vih);
                            Marshal.PtrToStructure(pSC, vsc);

                            int fps = (int)(10000000.0 / vsc.MaxFrameInterval);
                            if ((minFps < 0) || (minFps > fps))
                                minFps = fps;

                            fps = (int)(10000000.0 / vsc.MinFrameInterval);
                            if ((maxFps < 0) || (maxFps < fps))
                                maxFps = fps;

                            string capline = String.Format("{0} x {1}, min fps {2:0.}, max fps {3:0.}, {4}",
                                    vih.BmiHeader.Width, vih.BmiHeader.Height, 10000000.0 / vsc.MaxFrameInterval, 10000000.0 / vsc.MinFrameInterval, formatName);

                            if ((vih.BmiHeader.Width == currentWidth) &&
                                (vih.BmiHeader.Height == currentHeight) &&
                                (mt.subType == currentSubType))
                            {
                                videoFormatIndex = comboBoxFormats.Items.Count;
                            }

                            ComboboxItem item = new ComboboxItem();
                            item.Text = capline;
                            item.Value = i;

                            comboBoxFormats.Items.Add(item);
                        }
                    }

                    DsUtils.FreeAMMediaType(mt);
                }

                if (videoFormatIndex >= 0)
                    comboBoxFormats.SelectedIndex = videoFormatIndex;

                if ((minFps >= 0) && (maxFps >= 0))
                {
                    for (int i = minFps; i <= maxFps; i++)
                    {
                        ComboboxItem item = new ComboboxItem();
                        item.Text = i.ToString();
                        item.Value = i;

                        comboBoxFrameRate.Items.Add(item);

                        if (currentFps == i)
                            comboBoxFrameRate.SelectedIndex = comboBoxFrameRate.Items.Count - 1;
                    }
                }
            }
            finally
            {
                Marshal.FreeHGlobal(pSC);
            }
        }
    }
}

File: windows/net/samples.vb/CaptureDS/VideoCapturePropertiesForm.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Text
Imports System.Windows.Forms
Imports DirectShowLib
Imports System.Runtime.InteropServices
Imports System.Runtime.InteropServices.ComTypes

Namespace CaptureDS
	Partial Public Class VideoCapturePropertiesForm
		Inherits Form
		Public Sub New()
			InitializeComponent()
		End Sub

		Private Sub SetFormat(ByVal formatIndex As Integer, ByVal frameRate As Integer)
			Dim capsCount, capSize As Integer
			Dim hr As Integer = VideoConfig.GetNumberOfCapabilities(capsCount, capSize)
			DsError.ThrowExceptionForHR(hr)

			Dim pSC As IntPtr = Marshal.AllocHGlobal(capSize)
			Dim mt As AMMediaType = Nothing

			Try
				Dim vih As New VideoInfoHeader()

				hr = VideoConfig.GetStreamCaps(formatIndex, mt, pSC)
				DsError.ThrowExceptionForHR(hr)

				If frameRate > 0 Then
					Marshal.PtrToStructure(mt.formatPtr, vih)
					vih.AvgTimePerFrame = CLng(Fix(10000000.0 / frameRate))
					Marshal.StructureToPtr(vih, mt.formatPtr, False)
				End If

				hr = VideoConfig.SetFormat(mt)
				DsError.ThrowExceptionForHR(hr)
			Finally
				DsUtils.FreeAMMediaType(mt)
				Marshal.FreeHGlobal(pSC)
			End Try
		End Sub

		Private Sub buttonOK_Click(ByVal sender As Object, ByVal e As EventArgs) Handles buttonOK.Click
			Dim itemFormat As ComboboxItem = TryCast(comboBoxFormats.SelectedItem, ComboboxItem)
			If itemFormat IsNot Nothing Then
				Dim frameRate As Integer = -1
				Dim itemFPS As ComboboxItem = TryCast(comboBoxFrameRate.SelectedItem, ComboboxItem)
				If Nothing IsNot itemFPS Then
					frameRate = itemFPS.Value
				End If

				Try
					SetFormat(itemFormat.Value, frameRate)
				Catch e1 As Exception
					MessageBox.Show("Failed to set the selected video format.")
				End Try
			End If

			DialogResult = System.Windows.Forms.DialogResult.OK
		End Sub

		Private Sub buttonCancel_Click(ByVal sender As Object, ByVal e As EventArgs) Handles buttonCancel.Click
			DialogResult = System.Windows.Forms.DialogResult.Cancel
		End Sub

		Private privateVideoConfig As IAMStreamConfig
		Public Property VideoConfig() As IAMStreamConfig
			Get
				Return privateVideoConfig
			End Get
			Set(ByVal value As IAMStreamConfig)
				privateVideoConfig = value
			End Set
		End Property

		Private Structure FormatEntry
			Public videoSubType As Guid
			Public Name As String

			Public Sub New(ByVal videoSubType As Guid, ByVal name As String)
				Me.videoSubType = videoSubType
				Me.Name = name
			End Sub
		End Structure

		Private Shared FormatsTab() As FormatEntry = { New FormatEntry(MediaSubType.MJPG, "MJPG"), New FormatEntry(MediaSubType.RGB24, "RGB24"), New FormatEntry(MediaSubType.ARGB32, "ARGB32"), New FormatEntry(MediaSubType.RGB32, "RGB32"), New FormatEntry(MediaSubType.RGB565, "RGB565"), New FormatEntry(MediaSubType.ARGB1555, "ARGB1555"), New FormatEntry(MediaSubType.RGB555, "RGB555"), New FormatEntry(MediaSubType.ARGB4444, "ARGB4444"), New FormatEntry(MediaSubType.YV12, "YV12"), New FormatEntry(MediaSubType.I420, "I420"), New FormatEntry(MediaSubType.IYUV, "IYUV"), New FormatEntry(MediaSubType.YUY2, "YUY2"), New FormatEntry(MediaSubType.NV12, "NV12"), New FormatEntry(MediaSubType.UYVY, "UYVY"), New FormatEntry(MediaSubType.Y411, "Y411"), New FormatEntry(MediaSubType.Y41P, "Y41P"), New FormatEntry(MediaSubType.YVU9, "YVU9") }

		Private Function GetSubtypeString(ByVal videoSubtype As Guid) As String
			For i As Integer = 0 To FormatsTab.Length - 1
				If FormatsTab(i).videoSubType = videoSubtype Then
					Return FormatsTab(i).Name
				End If
			Next i

			Return Nothing
		End Function

		Public Class ComboboxItem
			Private privateText As String
			Public Property Text() As String
				Get
					Return privateText
				End Get
				Set(ByVal value As String)
					privateText = value
				End Set
			End Property
			Private privateValue As Integer
			Public Property Value() As Integer
				Get
					Return privateValue
				End Get
				Set(ByVal value As Integer)
					privateValue = value
				End Set
			End Property

			Public Overrides Function ToString() As String
				Return Text
			End Function
		End Class

		Private Sub VideoCapturePropertiesForm_Load(ByVal sender As Object, ByVal e As EventArgs) Handles MyBase.Load
			Try
				InitFormatsList()
			Catch e1 As Exception
				MessageBox.Show("Failed to set init the formats list.")
			End Try
		End Sub

		Private Sub InitFormatsList()
			Dim hr As Integer = 0

			If Nothing Is VideoConfig Then
				Throw New Exception("VideoConfig not set")
			End If

			' enumerate the capabilities of the video capture device
			Dim capsCount, capSize As Integer
			hr = VideoConfig.GetNumberOfCapabilities(capsCount, capSize)
			DsError.ThrowExceptionForHR(hr)

			Dim vih As New VideoInfoHeader()
			Dim vsc As New VideoStreamConfigCaps()
			Dim pSC As IntPtr = Marshal.AllocHGlobal(capSize)

			Try
				Dim videoFormatIndex As Integer = -1
				Dim minFps As Integer = -1
				Dim maxFps As Integer = -1

				Dim currentWidth As Integer = 0
				Dim currentHeight As Integer = 0
				Dim currentSubType As Guid
				Dim currentFps As Integer = 0

                Dim mt1 As AMMediaType = Nothing
                hr = VideoConfig.GetFormat(mt1)
                Marshal.PtrToStructure(mt1.formatPtr, vih)
                DsError.ThrowExceptionForHR(hr)

                currentFps = CInt(Fix(10000000.0 / vih.AvgTimePerFrame))
                currentWidth = vih.BmiHeader.Width
                currentHeight = vih.BmiHeader.Height
                currentSubType = mt1.subType

                DsUtils.FreeAMMediaType(mt1)

				For i As Integer = 0 To capsCount - 1
					Dim mt As AMMediaType = Nothing

					' the video format is described in AMMediaType and VideoStreamConfigCaps
					hr = VideoConfig.GetStreamCaps(i, mt, pSC)
					DsError.ThrowExceptionForHR(hr)

					If mt.formatType = DirectShowLib.FormatType.VideoInfo Then
						Dim formatName As String = GetSubtypeString(mt.subType)

						If (Not String.IsNullOrEmpty(formatName)) Then
							' copy the unmanaged structures to managed in order to check the format
							Marshal.PtrToStructure(mt.formatPtr, vih)
							Marshal.PtrToStructure(pSC, vsc)

							Dim fps As Integer = CInt(Fix(10000000.0 / vsc.MaxFrameInterval))
							If (minFps < 0) OrElse (minFps > fps) Then
								minFps = fps
							End If

							fps = CInt(Fix(10000000.0 / vsc.MinFrameInterval))
							If (maxFps < 0) OrElse (maxFps < fps) Then
								maxFps = fps
							End If

							Dim capline As String = String.Format("{0} x {1}, min fps {2:0.}, max fps {3:0.}, {4}", vih.BmiHeader.Width, vih.BmiHeader.Height, 10000000.0 / vsc.MaxFrameInterval, 10000000.0 / vsc.MinFrameInterval, formatName)

							If (vih.BmiHeader.Width = currentWidth) AndAlso (vih.BmiHeader.Height = currentHeight) AndAlso (mt.subType = currentSubType) Then
								videoFormatIndex = comboBoxFormats.Items.Count
							End If

							Dim item As New ComboboxItem()
							item.Text = capline
							item.Value = i

							comboBoxFormats.Items.Add(item)
						End If
					End If

					DsUtils.FreeAMMediaType(mt)
				Next i

				If videoFormatIndex >= 0 Then
					comboBoxFormats.SelectedIndex = videoFormatIndex
				End If

				If (minFps >= 0) AndAlso (maxFps >= 0) Then
					For i As Integer = minFps To maxFps
						Dim item As New ComboboxItem()
						item.Text = i.ToString()
						item.Value = i

						comboBoxFrameRate.Items.Add(item)

						If currentFps = i Then
							comboBoxFrameRate.SelectedIndex = comboBoxFrameRate.Items.Count - 1
						End If
					Next i
				End If
			Finally
				Marshal.FreeHGlobal(pSC)
			End Try
		End Sub
	End Class
End Namespace
